cpp/src/arrow/compute/kernels/scalar_cast_nested.cc

```diff
@@ -150,6 +150,78 @@ void AddListCast(CastFunction* func) {
   DCHECK_OK(func->AddKernel(SrcType::type_id, std::move(kernel)));
 }
 
+struct CastStruct {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    const CastOptions& options = CastState::Get(ctx);
+    const auto in_field_count =
+        checked_cast<const StructType&>(*batch[0].type()).num_fields();
+    const auto out_field_count =
+        checked_cast<const StructType&>(*out->type()).num_fields();
+
+    if (in_field_count != out_field_count) {
+      return Status::TypeError(
+          "struct field sizes do not match: ", batch[0].type()->ToString(), " ", " and ",
+          out->type()->ToString());
+    }
+
+    for (int64_t i = 0; i < in_field_count; ++i) {
+      const auto in_field_name =
+          checked_cast<const StructType&>(*batch[0].type()).field(i)->name();
+      const auto out_field_name =
+          checked_cast<const StructType&>(*out->type()).field(i)->name();
+      if (in_field_name != out_field_name) {
+        return Status::TypeError(
+            "struct field names do not match: ", batch[0].type()->ToString(), " ",
+            out->type()->ToString());
+      }
+    }
+
+    if (out->kind() == Datum::SCALAR) {
+      const auto& in_scalar = checked_cast<const StructScalar&>(*batch[0].scalar());
+      auto out_scalar = checked_cast<StructScalar*>(out->scalar().get());
+
+      DCHECK(!out_scalar->is_valid);
+      if (in_scalar.is_valid) {
+        for (int64_t i = 0; i < in_field_count; i++) {
+          auto values = in_scalar.value[i];
+          auto target_type = out->type()->field(i)->type();
+          ARROW_ASSIGN_OR_RAISE(Datum cast_values,
+                                Cast(values, target_type, options, ctx->exec_context()));
+          DCHECK_EQ(Datum::SCALAR, cast_values.kind());
+          out_scalar->value.push_back(cast_values.scalar());
+        }
+        out_scalar->is_valid = true;
+      }
+      return Status::OK();
+    }
+
+    const ArrayData& in_array = *batch[0].array();
+    ArrayData* out_array = out->mutable_array();
+
+    for (int64_t i = 0; i < in_field_count; ++i) {
+      auto values = in_array.child_data[i];
```

Comment:
Not complete but hope this is headed in the right direction. Think I just need to figure out how the slices are handled within this loop

Reply:
This looks right. (Absolute offset = child offset + parent offset)